Framework/Core/src/CompletionPolicyHelpers.cxx

```diff
@@ -21,6 +22,51 @@ namespace o2
 namespace framework
 {
 
+CompletionPolicy CompletionPolicyHelpers::defineByNameOrigin(std::string const& name, std::string const& origin, CompletionPolicy::CompletionOp op)
+{
+  auto matcher = [name](DeviceSpec const& device) -> bool {
+    return std::regex_match(device.name.begin(), device.name.end(), std::regex(name));
+  };
+
+  auto originReceived = std::make_shared<std::vector<uint64_t>>();
+
+  auto callback = [originReceived, origin, op](CompletionPolicy::InputSet inputRefs) -> CompletionPolicy::CompletionOp {
+    // update list of the start times of inputs with origin @origin
+    for (auto& ref : inputRefs) {
+      if (ref.header != nullptr) {
+        auto header = CompletionPolicyHelpers::getHeader<o2::header::DataHeader>(ref);
+        if (header->dataOrigin.str == origin) {
+          auto startTime = DataRefUtils::getHeader<DataProcessingHeader*>(ref)->startTime;
+          auto it = std::find(originReceived->begin(), originReceived->end(), startTime);
+          if (it == originReceived->end()) {
+            originReceived->emplace_back(startTime);
+          }
+        }
+      }
+    }
+
+    // find out if all inputs which are not of origin @origin have a corresponding entry in originReceived
```

Comment:
This is fine, however I think it's in general more readable to have for loops where you continue the loop, rather than many nested if statements. E.g.

```C++
for (auto& ref : inputRefs) {
  if (ref.header == nullptr) {
    continue;
  }
  
  auto header = CompletionPolicyHelpers::getHeader<o2::header::DataHeader>(ref);
  if (header->dataOrigin.str != origin) {
    continue;
  }
  auto startTime = DataRefUtils::getHeader<DataProcessingHeader*>(ref)->startTime;
  auto it = std::find(originReceived->begin(), originReceived->end(), startTime);
  if (it != originReceived->end()) {
    continue;
  }
  originReceived->emplace_back(startTime);
}
```